src/Storages/MergeTree/MergeTreeDataPartCloner.cpp

```diff
@@ -0,0 +1,394 @@
+#include <Interpreters/MergeTreeTransaction.h>
+#include <Storages/MergeTree/MergeTreeDataPartBuilder.h>
+#include <Storages/MergeTree/MergeTreeDataPartCloner.h>
+#include <Common/escapeForFileName.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+Poco::Logger * MergeTreeDataPartCloner::log = &Poco::Logger::get("MergeTreeDataPartCloner1");
+
+namespace DistinctPartitionExpression
+{
+
+    static std::unique_ptr<WriteBufferFromFileBase> updatePartitionFile(
+        const MergeTreeData & merge_tree_data,
+        const MergeTreePartition & partition,
+        const MergeTreeDataPartCloner::MutableDataPartPtr & dst_part,
+        IDataPartStorage & storage
+    )
+    {
+        storage.removeFile("partition.dat");
+        // Leverage already implemented MergeTreePartition::store to create & store partition.dat.
+        // Checksum is re-calculated later.
+        return partition.store(merge_tree_data, storage, dst_part->checksums);
+    }
+
+    static void deleteMinMaxFiles(
+        const MergeTreeData & merge_tree_data,
+        IDataPartStorage & storage,
+        const StorageMetadataPtr & metadata_snapshot
+    )
+    {
+        for (const auto & column_name : merge_tree_data.getMinMaxColumnsNames(metadata_snapshot->partition_key))
+        {
+            auto file = "minmax_" + escapeForFileName(column_name) + ".idx";
+            storage.removeFile(file);
+        }
+    }
+
+    static IMergeTreeDataPart::MinMaxIndex::WrittenFiles updateMinMaxFiles(
+        const MergeTreeData & merge_tree_data,
+        const MergeTreeDataPartCloner::MutableDataPartPtr & dst_part,
+        IDataPartStorage & storage,
+        const StorageMetadataPtr & metadata_snapshot
+    )
+    {
+        deleteMinMaxFiles(merge_tree_data, storage, metadata_snapshot);
+
+        return dst_part->minmax_idx->store(merge_tree_data, storage, dst_part->checksums);
```

Comment:
All of these operations are happening in a temporary directory.

```
arthur@arthur:~/ClickHouse/programs/server/data/default/destination$ ls
detached  format_version.txt  tmp_replace_from_14670_1_1_0
```

Once operation completes, it gets renamed in a transactional manner. See https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageMergeTree.cpp#L2067.

```
arthur@arthur:~/ClickHouse/programs/server/data/default/destination$ ls
14670_1_1_0  detached  format_version.txt
```



If an exception is thrown, operation will be aborted and it will leave a `tmp` directory with some files there. I assume this is cleaned up at some point.